File: controllers/project.go
package controllers

import (
	"fmt"
	"math"
	"net/http"
	"path/filepath"
	"strconv"
	"strings"

	"kasen/errs"
	"kasen/server"
	"kasen/services"
)

func Cover(c *server.Context) {
	slug := c.Param("slug")
	fileName := c.Param("fileName")

	if len(slug) == 0 || len(fileName) == 0 {
		c.Status(http.StatusBadRequest)
		return
	}

	str := strings.TrimPrefix(c.Param("width"), "/")
	width, _ := strconv.Atoi(strings.TrimSuffix(str, filepath.Ext(str)))

	c.Header("Cache-Control", "public, max-age=300")
	services.ServeCover(slug, fileName, int(width), c.Writer, c.Request)
}

type ProjectParams struct {
	ID   int64  `uri:"id"`
	Slug string `uri:"slug"`
}

func Project(c *server.Context) {
	params := &ProjectParams{}
	c.BindUri(params)

	page, _ := strconv.Atoi(c.Query("page"))
	if page <= 0 {
		page = 1
	}

	if params.ID <= 0 {
		c.SetData("error", errs.ErrProjectNotFound)
		c.HTML(http.StatusInternalServerError, "error.html")
		return
	}

	templateName := "project.html"
	if c.TryCache(templateName) {
		return
	}

	o := services.GetProjectOptions{
		Preloads: []string{
			services.ProjectRels.Artists,
			services.ProjectRels.Authors,
			services.ProjectRels.Cover,
			services.ProjectRels.Tags,
			services.ProjectRels.Statistic,
		},
	}

	pResult := services.GetProject(params.ID, o)
	if pResult.Err != nil {
		c.SetData("error", pResult.Err)
		c.HTML(http.StatusInternalServerError, "error.html")
		return
	}

	project := pResult.Project
	if project.Slug != strings.TrimPrefix(params.Slug, "/") {
		c.Redirect(http.StatusFound, fmt.Sprintf("/projects/%d/%s", project.ID, project.Slug))
		return
	}

	limit := 20
	cResult := services.GetChapters(services.GetChaptersOptions{
		ProjectID: pResult.Project.ID,

		Limit:  limit,
		Offset: limit * (page - 1),

		Preloads: []string{
			services.ChapterRels.Uploader,
			services.ChapterRels.ScanlationGroups,
			services.ChapterRels.Statistic,
		},
		GetThumbnail: true,
	})
	if cResult.Err != nil {
		c.SetData("error", cResult.Err)
		c.HTML(http.StatusInternalServerError, "error.html")
		return
	}

	services.IncreaseViewCount(project.Stats, c.ClientIP())
	project.Chapters = cResult.Chapters

	for _, c := range project.Chapters {
		c.Stats = services.GetChapterStats(c.ID).Stats
	}

	c.SetData("project", project)
	c.SetData("chapters", cResult.Chapters)
	c.SetData("totalChapters", cResult.Total)

	totalPages := int(math.Ceil(float64(cResult.Total) / float64(limit)))
	c.SetData("pagination", services.CreatePagination(page, totalPages))

	c.Cache(http.StatusOK, templateName)
}

type ProjectsQueries struct {
	Title                 string   `form:"title"`
	ProjectStatus         []string `form:"projectStatus"`
	SeriesStatus          []string `form:"seriesStatus"`
	Demographic           []string `form:"demographic"`
	Rating                []string `form:"rating"`
	ExcludedProjectStatus []string `form:"excludeProjectStatus"`
	ExcludedSeriesStatus  []string `form:"excludeSeriesStatus"`
	ExcludedDemographic   []string `form:"excludeDemographic"`
	ExcludedRating        []string `form:"excludeRating"`
	Artists               []string `form:"artist"`
	Authors               []string `form:"author"`
	Tags                  []string `form:"tag"`
	ExcludedTags          []string `form:"excludeTag"`
	Page                  int      `form:"page"`
	Sort                  string   `form:"sort"`
	Order                 string   `form:"order"`
}

func Projects(c *server.Context) {
	q := &ProjectsQueries{}
	c.BindQuery(q)

	templateName := "projects.html"
	if c.TryCache(templateName) {
		return
	}

	c.SetData("queries", q)
	c.SetData("hasQueries",
		len(q.Title) > 0 ||
			len(q.ProjectStatus) > 0 ||
			len(q.SeriesStatus) > 0 ||
			len(q.Demographic) > 0 ||
			len(q.Rating) > 0 ||
			len(q.ExcludedProjectStatus) > 0 ||
			len(q.ExcludedSeriesStatus) > 0 ||
			len(q.ExcludedDemographic) > 0 ||
			len(q.ExcludedRating) > 0 ||
			len(q.Artists) > 0 ||
			len(q.Authors) > 0 ||
			len(q.Tags) > 0 ||
			len(q.ExcludedTags) > 0)

	limit := 24
	result := services.GetProjects(services.GetProjectsOptions{
		Title:                 q.Title,
		ProjectStatus:         q.ProjectStatus,
		SeriesStatus:          q.SeriesStatus,
		Demographic:           q.Demographic,
		Rating:                q.Rating,
		ExcludedProjectStatus: q.ExcludedProjectStatus,
		ExcludedSeriesStatus:  q.ExcludedSeriesStatus,
		ExcludedDemographic:   q.ExcludedDemographic,
		ExcludedRating:        q.ExcludedRating,
		Artists:               q.Artists,
		Authors:               q.Authors,
		Tags:                  q.Tags,
		ExcludedTags:          q.ExcludedTags,
		Limit:                 limit,
		Offset:                limit * (q.Page - 1),
		Sort:                  q.Sort,
		Order:                 q.Order,

		Preloads: []string{
			services.ProjectRels.Cover,
			services.ProjectRels.Tags,
		},
	})
	c.SetData("projects", result.Projects)
	c.SetData("total", result.Total)

	totalPages := int(math.Ceil(float64(result.Total) / float64(limit)))
	c.SetData("pagination", services.CreatePagination(q.Page, totalPages))

	tags, _ := services.GetTags()
	c.SetData("tags", tags)

	c.Cache(http.StatusOK, templateName)
}

File: web/src/api/project.ts
import { Demographic, ProjectStatus, Rating, SeriesStatus } from "../constants";
import { Order, ProjectPreloads, ProjectSort } from "./enums";
import SendRequest from "./xhr";

interface CheckProjectOptions {
  id?: number;
  slug?: string;
  title?: string;
}

export const CheckProjectExists = ({ id, slug, title }: CheckProjectOptions) =>
  SendRequest<{ id: string; slug: string }>(
    "GET",
    `/api/project/exists?id=${id || ""}&slug=${slug || ""}&title=${title || ""}`
  );

export const CreateProject = (draft: ProjectDraft) =>
  SendRequest<Project>("POST", `/api/project`, JSON.stringify(draft));

export const DeleteProject = (id: number) => SendRequest("DELETE", `/api/project/${id}`);

interface GetProjectOptions {
  preloads?: ProjectPreloads[];
  includesDrafts?: boolean;
}

export const GetProject = (id: number, o: GetProjectOptions) => {
  const searchParams = new URLSearchParams();

  if (o.preloads?.length) o.preloads.forEach(preload => searchParams.append("preload", preload));
  if (o.includesDrafts) searchParams.set("includesDrafts", "true");

  let url = `/api/project/${id}`;
  const params = searchParams.toString();
  if (params.length) url += `?${params}`;

  return SendRequest<Project>("GET", url);
};

export const GetProjectMd = (id: string) => SendRequest<ProjectDraft>("GET", `/api/md/project/${id}`);

interface GetProjectsOptions {
  title?: string;

  projectStatus?: ProjectStatus;
  seriesStatus?: SeriesStatus;
  demographic?: Demographic;
  rating?: Rating;

  excludedProjectStatus?: string[];
  excludedSeriesStatus?: string[];
  excludedDemographic?: string[];
  excludedRating?: string[];

  artists?: string[];
  authors?: string[];
  tags?: string[];

  limit?: number;
  offset?: number;
  preloads?: ProjectPreloads[];
  sort?: ProjectSort;
  order?: Order;
  includesDrafts?: boolean;
}

interface GetProjectsResult {
  data?: Project[];
  total?: number;
}

export const GetProjects = (o: GetProjectsOptions) => {
  const searchParams = new URLSearchParams();

  if (o.title) searchParams.set("title", o.title);
  if (o.projectStatus) searchParams.set("projectStatus", o.projectStatus);
  if (o.demographic) searchParams.set("demographic", o.demographic);
  if (o.rating) searchParams.set("rating", o.rating);

  if (o.excludedProjectStatus?.length)
    o.excludedProjectStatus.forEach(v => searchParams.append("excludeProjectStatus", v));
  if (o.excludedSeriesStatus?.length)
    o.excludedSeriesStatus.forEach(v => searchParams.append("excludeSeriesStatus", v));
  if (o.excludedDemographic?.length) o.excludedDemographic.forEach(v => searchParams.append("excludeDemographic", v));
  if (o.excludedRating?.length) o.excludedRating.forEach(v => searchParams.append("excludeRating", v));

  if (o.artists?.length) o.artists.forEach(v => searchParams.append("artist", v));
  if (o.authors?.length) o.authors.forEach(v => searchParams.append("author", v));
  if (o.tags?.length) o.tags.forEach(v => searchParams.append("tag", v));

  if (o.limit > 0) searchParams.set("limit", o.limit.toString());
  if (o.offset >= 0) searchParams.set("offset", o.offset.toString());
  if (o.sort) searchParams.set("sort", o.sort);
  if (o.order) searchParams.set("order", o.order);

  if (o.preloads?.length) o.preloads.forEach(rel => searchParams.append("preload", rel));
  if (o.includesDrafts) searchParams.set("includesDrafts", "true");

  let url = "/api/project";
  const params = searchParams.toString();
  if (params.length) url += `?${params}`;

  return SendRequest<GetProjectsResult>("GET", url);
};

export const LockProject = (id: number) => SendRequest<Project>("PATCH", `/api/project/${id}/lock`);

export const PublishProject = (id: number) => SendRequest<Project>("PATCH", `/api/project/${id}/publish`);

export const UnlockProject = (id: number) => SendRequest<Project>("PATCH", `/api/project/${id}/unlock`);

export const UnpublishProject = (id: number) => SendRequest<Project>("PATCH", `/api/project/${id}/unpublish`);

export const UpdateProject = (id: number, draft: ProjectDraft) =>
  SendRequest<Project>("PATCH", `/api/project/${id}`, JSON.stringify(draft));
